src/register_target.cpp

```diff
@@ -33,42 +34,88 @@ inline namespace MIGRAPHX_INLINE_NS {
 void store_target_lib(const dynamic_loader& lib)
 {
     static std::vector<dynamic_loader> target_loader;
+    static std::mutex mutex;
+    std::unique_lock<std::mutex> lock(mutex);
+
     target_loader.emplace_back(lib);
 }
 
+/**
+ * Returns a singleton map of targets and names.
+ */
 std::unordered_map<std::string, target>& target_map()
 {
     static std::unordered_map<std::string, target> m; // NOLINT
     return m;
 }
 
+/**
+ * Returns a singleton mutex used by the various register_target methods.
+ */
+std::mutex& target_mutex()
+{
+    static std::mutex m; // NOLINT
+    return m;
+}
+
 void register_target_init() { (void)target_map(); }
 
 void unregister_target(const std::string& name)
 {
+    std::unique_lock<std::mutex> lock(target_mutex());
     assert(target_map().count(name));
     target_map().erase(name);
 }
 
-void register_target(const target& t) { target_map()[t.name()] = t; }
+/**
+ * Insert a target name in the target_map; thread safe.
+ */
+void register_target(const target& t)
+{
+    std::unique_lock<std::mutex> lock(target_mutex());
+    target_map()[t.name()] = t;
+}
+
+/**
+ * Search for a target by name in the target_map; thread-safe.
+ */
+migraphx::optional<target> find_target(const std::string& name)
+{
+    // search for match or return none
+    std::unique_lock<std::mutex> lock(target_mutex());
+    const auto it = target_map().find(name);
+
+    if(it == target_map().end())
+        return nullopt;
+    return it->second;
+}
 
+/**
+ * Get a target by name.  Load target library and register target if needed.
+ * Thread safe.
+ */
 target make_target(const std::string& name)
 {
-    if(not contains(target_map(), name))
+    //   no lock required here
+    auto t = find_target(name);
+    if(t == nullopt)
     {
         std::string target_name = "libmigraphx_" + name + ".so";
+        // register_target is called by this
         store_target_lib(dynamic_loader(target_name));
+        t = find_target(name);
     }
-    const auto it = target_map().find(name);
-    if(it == target_map().end())
-    {
-        MIGRAPHX_THROW("Requested target '" + name + "' is not loaded or not supported");
-    }
-    return it->second;
+    // at this point we should always have a target
```

Comment:
Why is it guaranteed to have a target at this point?

We have Three cases. 
1. Target is already registered. 
2. Target is not registered but can be dynamically loaded. 
3. Target requested is not supported by MIGraphX e.g. some target named "TPU". TPU is not registered with MIGraphX and it will not be found or can not be loaded dynamically either.

Reply:
Agreed, we should check `if (t == nullopt)` a second time and throw.

Reply:
> Why is it guaranteed to have a target at this point? Let's say Target requested is not supported by MIGraphX e.g. some target named "TPU". TPU is not registered with MIGraphX and it will not be found or can not be loaded dynamically either.

Because the `dynamic_loader` constructor will throw an exception.  I added a line in the test for this.

Reply:
> Agreed, we should check `if (t == nullopt)` a second time and throw.

That's redundant since `find_target` contains that check.